src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyManager.cs

```diff
@@ -0,0 +1,228 @@
+                }
+            }
+            else if (CanDecreaseConcurrency(functionConcurrencyStatus))
+            {
+                functionConcurrencyStatus.DecreaseConcurrency();
+            }
+
+            _logger.HostConcurrencyStatus(functionId, functionConcurrencyStatus.CurrentConcurrency, functionConcurrencyStatus.OutstandingInvocations);
+
+            return functionConcurrencyStatus;
+        }
+
+        internal virtual HostConcurrencySnapshot GetSnapshot()
+        {
+            var functionSnapshots = ConcurrencyStatuses.Values.ToDictionary(p => p.FunctionId, q => new FunctionConcurrencySnapshot { Concurrency = q.CurrentConcurrency }, StringComparer.OrdinalIgnoreCase);
+            var hostSnapshot = new HostConcurrencySnapshot
+            {
+                Timestamp = DateTime.UtcNow,
+                NumberOfCores = EffectiveCoresCount,
+                FunctionSnapshots = functionSnapshots
+            };
+            return hostSnapshot;
+        }
+
+        internal virtual void ApplySnapshot(HostConcurrencySnapshot hostSnapshot)
+        {
+            if (hostSnapshot == null)
+            {
+                throw new ArgumentNullException(nameof(hostSnapshot));
+            }
+
+            if (hostSnapshot.FunctionSnapshots != null)
+            {
+                foreach (var functionSnapshot in hostSnapshot.FunctionSnapshots)
+                {
+                    int concurrency = GetCoreAdjustedConcurrency(functionSnapshot.Value.Concurrency, hostSnapshot.NumberOfCores, EffectiveCoresCount);
+                    functionSnapshot.Value.Concurrency = concurrency;
+
+                    // Since we may be initializing for functions that haven't run yet, if the snapshot contains
+                    // stale functions, they'll be added. When we write snapshots, we prune stale entries though.
+                    var concurrencyStatus = GetFunctionConcurrencyStatus(functionSnapshot.Key);
+                    _logger.LogInformation($"Applying status snapshot for function {functionSnapshot.Key} (Concurrency: {functionSnapshot.Value.Concurrency})");
+                    concurrencyStatus.ApplySnapshot(functionSnapshot.Value);
+                }
+            }
+        }
+
+        internal static int GetCoreAdjustedConcurrency(int concurrency, int otherCores, int cores)
+        {
+            if (cores != otherCores)
+            {
+                // To allow for variance across machines, we compute the effective concurrency
+                // based on number of cores. When running in an App Service plan, all instances will have
+                // the same VM specs. When runing in the Consumption plan, VMs may differ. In the latter case,
+                // if the snapshot ere taken on a VM with a different core count than ours, the adjusted
+                // concurency we compute may not be optimal, but it's a starting point that we'll dynamically
```

Comment:
> When runing in the Consumption plan, VMs may differ.

I'm trying to understand this comment. As far as I can tell, the `GetEffectiveCoresCount` method already accounted for consumption plan VMs by returning a single core, regardless of the actual VM specs. What dynamic adjustment are you referring to for Consumption plan VMs?

Reply:
GetEffectiveCoresCount returns the actual Environment.ProcessorCount on consumption VMSS. That's the variability I'm trying to account for here. Similarly, if a snapshot exists, and the ASP was scaled up, we could see differences as well.